Mathlib/Algebra/Order/Hom/Quantale.lean

```diff
@@ -0,0 +1,230 @@
+def toQuantaleHom [MulHomClass F α β] [sSupHomClass F α β] (f : F) :
+    α →ₙ*q β := { (f : α →ₙ* β) with map_sSup' := sSupHomClass.map_sSup f }
+
+/-- Any type satisfying `QuantaleHomClass` can be cast into `QuantaleHom` via
+  `QuantaleHomClass.toQuantaleHom`. -/
+@[to_additive "Any type satisfying `AddQuantaleHomClass` can be cast into `AddQuantaleHom` via
+  `AddQuantaleHomClass.toAddQuantaleHom`."]
+instance [MulHomClass F α β] [sSupHomClass F α β] : CoeTC F (α →ₙ*q β) :=
+  ⟨toQuantaleHom⟩
+
+end Quantale
+
+namespace QuantaleHom
+
+/- Now, to be honest, I don't really understand which theorems below are necessary
+and which ones should already be there in some other way. While proving some,
+I got the impression I was overlooking something very basic.
+
+My main confusion at the moment, is how `HomClass`es are being used, and how
+I get automatic results about `Hom`s from the `HomClass` theorems. See for
+example the `toOrderHom` theorem below.
+
+Which ones should I have, which ones do I not need, and why??
+-/
+
+variable [Semigroup α] [Semigroup β] [CompleteLattice α] [CompleteLattice β] {f g : α →ₙ*q β}
+
+@[to_additive]
+instance : FunLike (α →ₙ*q β) α β where
+  coe f := f.toFun
+  coe_injective' f g h := by
+    obtain ⟨⟨ _ ⟩, _⟩ := f
+    obtain ⟨⟨ _ ⟩, _⟩ := g
+    congr
+
+@[to_additive]
+instance : MulHomClass (α →ₙ*q β) α β where
+  map_mul f _ _ := f.map_mul' _ _
+
+@[to_additive]
+instance : sSupHomClass (α →ₙ*q β) α β where
+  map_sSup f := f.map_sSup'
+
+-- Other lemmas should be accessed through the `FunLike` API
+@[to_additive (attr := ext)]
+theorem ext (h : ∀ a, f a = g a) : f = g :=
+  DFunLike.ext f g h
+
+@[to_additive]
+theorem toFun_eq_coe (f : α →ₙ*q β) : f.toFun = (f : α → β) :=
+  rfl
+
+@[to_additive (attr := simp)]
+theorem coe_mk (f : α →ₙ* β) (h) : (QuantaleHom.mk f h : α → β) = f :=
+  rfl
+
+@[to_additive (attr := simp)]
+theorem mk_coe (f : α →ₙ*q β) (h) : QuantaleHom.mk (f : α →ₙ* β) h = f := by
+  ext
+  rfl
+
+/-- Reinterpret a quantale homomorphism as a supsemilattice homomorphism-/
+@[to_additive "Reinterpret an additive quantale homomorphism as a supsemilattice homomorphism."]
+def toSupBotHom (f : α →ₙ*q β) : SupBotHom α β where
+  toFun := f
+  map_sup':= sSupHomClass.toSupBotHomClass.map_sup f
+  map_bot':= sSupHomClass.toSupBotHomClass.map_bot f
+
+@[to_additive (attr := simp)]
+theorem coe_SupBotHom (f : α →ₙ*q β) : ((f : SupBotHom α β) : α → β) = f :=
+  rfl
+
+/-- Reinterpret a quantale homomorphism as an order homomorphism. -/
+@[to_additive "Reinterpret an additive quantale homomorphism as an order homomorphism."]
+def toOrderHom (f : α →ₙ*q β) : α →o β where
+  toFun := f
+  monotone' := by
+    -- This should not be so elaborate, what am I overlooking? --
+    intro x y h
+    apply le_of_sup_eq
+    rw [← (f : SupBotHom α β).map_sup']
+    dsimp only [toFun_eq_coe]
+    congr
+    apply sup_of_le_right
+    exact h
```

Comment:
Do you actually need these definitions? They come automatically from `sSupHomClass.toSupBotHomClass` and alike